core/src/main/java/com/linecorp/armeria/internal/metric/PrometheusMetricRequestDecorator.java

```diff
@@ -0,0 +1,311 @@
+     *
+     * @param <I> Request type
+     * @param <O> Response type
+     * @param collectorRegistry Prometheus registry
+     * @param labels Labels names for metrics
+     * @param labelingFunction Must return a sorted map in natural order
+     * @return A service decorator function
+     */
+    public static <T extends MetricLabel<T>, I extends Request, O extends Response>
+    PrometheusMetricRequestDecorator<T, I, O> decorateService(
+            CollectorRegistry collectorRegistry,
+            T[] labels,
+            Function<RequestLog, Map<T, String>> labelingFunction) {
+        return new PrometheusMetricRequestDecorator<>(collectorRegistry,
+                                                      labels,
+                                                      labelingFunction,
+                                                      "armeria_server");
+    }
+
+    /**
+     * Returns a new {@link Client} decorator that tracks request stats using the Prometheus metrics  library.
+     *
+     * @param <I> Request type
+     * @param <O> Response type
+     * @param prometheusRegistry Prometheus registry
+     * @param labels Labels names for metrics
+     * @param labelingFunction map of labels to values
+     * @return A client decorator function
+     */
+    public static <T extends MetricLabel<T>, I extends Request, O extends Response>
+    PrometheusMetricRequestDecorator<T, I, O> decorateClient(
+            CollectorRegistry prometheusRegistry,
+            T[] labels,
+            Function<RequestLog, Map<T, String>> labelingFunction) {
+        return new PrometheusMetricRequestDecorator<>(prometheusRegistry,
+                                                      labels,
+                                                      labelingFunction,
+                                                      "armeria_client");
+    }
+
+    private final Function<RequestLog, Map<T, String>> labelingFunction;
+    private final MetricAdapter<T> metricAdapter;
+
+    /**
+     * Creates a new instance that decorates the specified {@link Service}.
+     */
+    private PrometheusMetricRequestDecorator(
+            CollectorRegistry collectorRegistry,
+            T[] metricLabels,
+            Function<RequestLog, Map<T, String>> labelingFunction,
+            String prefix) {
+        Objects.requireNonNull(collectorRegistry, "collectorRegistry");
+        Objects.requireNonNull(metricLabels, "metricLabels");
+        Objects.requireNonNull(labelingFunction, "labelingFunction");
+        Objects.requireNonNull(prefix, "prefix");
+
+        this.labelingFunction = labelingFunction;
+        this.metricAdapter = new MetricAdapter<>(collectorRegistry, metricLabels, prefix, HELP_STRING);
+    }
+
+    private <C extends RequestContext> O request(ThrowingBiFunction<C, I, O> function, C ctx, I req)
+            throws Exception {
+        ctx.log().addListener(
+                log -> {
```

Comment:
Realized I never responded to the point of using RequestLog. I realized that RequestLog is probably the only way to have access to both HTTP headers and RpcRequest at the same time, which is probably what the use case is.